Improving the quality of the API

Gisleburt opened the pull request:
**Done**:
- Provide BambooHTTPRequest injection and complete unit testing Bamboo API.
- Separate source files, implement PSR-4
- Implement PSR-1 coding standards (using squizlabs phpcs)
- Add other quality control mechanisms (phpmd, phpcpd)
- Added Travis CI
- Did a little tidy (might be BC breaking for some but I think it will be worth it, especially since this isn't in composer yet).

**Skipped:**
- Work out a way around use of curl
- Implement PSR-1 coding standards (using squizlabs phpcs)

Please let me know if any of these changes will prevent you accepting the PR so that I can skip them. 😉

Gisleburt commented:
Notes: 
- Tests will only work in PHP 5.6 (I can lower it to 5.5 by removing the variadic, but if we want to go lower it's a lot of extra work to remove the uses of `::class` naming).
- The injection method I use will only work in PHP 5.4. (traits)
- PHP 5.3 and 5.4 are no longer supported, PHP 5.5 is security supported up to next week (10th of July) then no longer supported at all.

jlesueur commented:
- composer.json should reflect the minimum php version, but I'm ok with upping it to 5.5. We would want to think about whether a minimum of 5.6 is ok. Debian et. al will be keeping some people on 5.5 beyond that PHP support date, and we have to be conscious of a pretty broad base.
- PSR-1 is fine, but PSR-2 is a no-go. I may be able to provide a standard you can use for phpcs. I'll get back to you on that.
- PSR-4 is good.
- For removing curl, can you share the goal there? are you looking at using guzzle, or some other library? We'd like to avoid dependencies as much as possible, so if the goal is testing, a thin wrapper might be an option?
- I'd love to hear more about how you'll be using traits for injection.

Haven't looked at the actual code yet, just wanted to answer questions first. I'm excited you're taking an interest in this.

jlesueur commented on tests/BambooAPITest.php:
```diff
@@ -0,0 +1,485 @@
+            $http,
+            $this->getObjectAttribute($bambooApi, 'httpHandler')
+        );
+    }
```
would it be worth anything to make these three separate tests?

Gisleburt replied:
Can do.

jlesueur commented on tests/BambooHTTPRequestTest.php:
```diff
@@ -0,0 +1,207 @@
+        $this->assertTrue($response->isError());
+
+        // ToDo: This is not an error either
+        $response->statusCode = 2001;
```
Isn't 2001 invalid? should we consider invalid status codes errors?

Gisleburt replied:
Sorry, for some reason I had it in mind that I'd seen a four digit status code, but I checked the spec and it is strictly 3 digits.

Gisleburt commented:
> composer.json should reflect the minimum php version, but I'm ok with upping it to 5.5. We would want to think about whether a minimum of 5.6 is ok. Debian et. al will be keeping some people on 5.5 beyond that PHP support date, and we have to be conscious of a pretty broad base.

I'll remove the variadic and set it to PHP 5.5

> PSR-1 is fine, but PSR-2 is a no-go. I may be able to provide a standard you can use for phpcs. I'll get back to you on that.

Sure, I'll set PSR-1 now, if you send me the alternate standards I'll add them in. Out of interest, why no PSR-2?

> PSR-4 is good.

Cool, I'll add it once the tests are better so we can be sure I didn't make a maess.

> For removing curl, can you share the goal there? are you looking at using guzzle, or some other library? We'd like to avoid dependencies as much as possible, so if the goal is testing, a thin wrapper might be an option?

I just wanted to make the method that uses it testable.

> I'd love to hear more about how you'll be using traits for injection.

It's nothing clever, it'll just be a getter/setter with a default returned if nothing has been set yet.

> Haven't looked at the actual code yet, just wanted to answer questions first. I'm excited you're taking an interest in this.

I'm not sure I can talk about it yet, certainly not in public, but if you drop me a DM on twitter (@gisleburt) then I'll let you know. 😉

Gisleburt commented:
That's all of the tests written. It should be 100% complete _except_ for the function that uses curl.

Gisleburt commented:
Worth noting there's three possible B/C breaks:
- Echo removed from `BambooAPI::addTimeOffHistoryFromRequest`
- `buildMultipart` has been moved inside `BambooHTTPRequest`
- `BAMBOOHR_MULTIPART_BOUNDARY` has been moved inside `BambooHTTPRequest`

Gisleburt commented:
Hmm, PHPUnit no longer supports PHP 5.5. I can fix that, I just need to change the why the mocks are generated I think

https://travis-ci.org/Gisleburt/bhr-api-php/jobs/142229609

Gisleburt commented:
Fixed, though PHP 7 screams warning at you for using `getMock` instead of `createMock`

jlesueur commented on tests/BambooAPITest.php:
```diff
@@ -176,310 +470,1747 @@ public function testPrepareKeyValues()
-        // ToDo:
-        $this->assertTrue((bool)'This can not be tested until BambooHTTPRequest is injected');
+        $companyDomain = 'CompanyDomain';
+        $employeeId = 42;
```
appropriate magic number :)

jlesueur commented on BambooHR/Injector/BambooHTTPRequestInjector.php:
```diff
@@ -0,0 +1,44 @@
+ * Allows injection of BambooHttpRequest with a default to fall back on
+ * @package BambooHR
+ */
+trait BambooHTTPRequestInjector
```
what advantage do you get from making this a trait vs. just adding these methods/member variables to the class?

jlesueur commented:
Here is the phpcs standard we use:
https://github.com/BambooHR/phpcs

You may notice there are some things in there that aren't compatible with psr-2 (spaces vs. tabs for instance, and placement of braces).

Gisleburt commented:
So, a year and a half later I was flicking through open PRs and saw this was a) still here, b) not in conflict, so I thought I'd finish it off. 

I've added your styles to phpcs and fixed what I can. However, although I agree with it, the `camelBackCase` rule will break backwards compatability. It wants you to change things like `BambooHTTPResponse::getContentXML` to `BambooHTTPResponse::getContentXml`, but you can't do that to public methods. I'll turn over the decision on what to do back to you. I hope you'll consider merging it anyway.

It was fun to work in PHP again, even if only very breifly. 😄

jlesueur commented on phpcs.xml:
```diff
@@ -0,0 +1,9 @@
+    <file>./src</file>
+    <file>./tests</file>
+
+    <rule ref="./vendor/bamboohr/phpcs"/>
```
should be "./vendor/bamboohr/phpcs/BambooHR"

jlesueur commented:
Thanks for the contribution. You've pushed this little library in a better direction.
